components/Components/URLSearchParamsTest.php

```diff
@@ -0,0 +1,494 @@
+    {
+        $params = new URLSearchParams('a=b&c=d');
+
+        self::assertSame($params->getAll('a'), ['b']);
+        self::assertSame($params->getAll('c'), ['d']);
+        self::assertSame($params->getAll('e'), []);
+
+        $params = new URLSearchParams('a=b&c=d&a=e');
+
+        self::assertSame($params->getAll('a'), ['b', 'e']);
+
+        $params = new URLSearchParams('=b&c=d');
+
+        self::assertSame($params->getAll(''), ['b']);
+
+        $params = new URLSearchParams('a=&c=d&a=e');
+
+        self::assertSame($params->getAll('a'), ['', 'e']);
+
+        $params = new URLSearchParams('a=1&a=2&a=3&a');
+
+        self::assertTrue($params->has('a'), 'Search params object has name "a"');
+
+        $matches = $params->getAll('a');
+
+        self::assertCount(4, $matches, 'Search params object has values for name "a"');
+        self::assertSame($matches, ['1', '2', '3', ''], 'Search params object has expected name "a" values');
+
+        $params->set('a', 'one');
+
+        self::assertSame($params->get('a'), 'one', 'Search params object has name "a" with value "one"');
+
+        $matches = $params->getAll('a');
+
+        self::assertCount(1, $matches, 'Search params object has values for name "a"');
+        self::assertSame($matches, ['one'], 'Search params object has expected name "a" values');
+    }
+
+    public function testSetMethod(): void
+    {
+        $params = new URLSearchParams('a=b&c=d');
+        $params->set('a', 'B');
+        self::assertSame($params->toString(), 'a=B&c=d');
+        $params = new URLSearchParams('a=b&c=d&a=e');
+        $params->set('a', 'B');
+        self::assertSame($params->toString(), 'a=B&c=d');
+        $params->set('e', 'f');
+        self::assertSame($params->toString(), 'a=B&c=d&e=f');
+
+        $params = new URLSearchParams('a=1&a=2&a=3');
+        self::assertTrue($params->has('a'), 'Search params object has name "a"');
+        self::assertSame($params->get('a'), '1', 'Search params object has name "a" with value "1"');
+        $params->set('first', 4);
+        self::assertTrue($params->has('a'), 'Search params object has name "a"');
+        self::assertSame($params->get('a'), '1', 'Search params object has name "a" with value "1"');
+        $params->set('a', 4);
+        self::assertTrue($params->has('a'), 'Search params object has name "a"');
+        self::assertSame($params->get('a'), '4', 'Search params object has name "a" with value "4"');
+    }
+
+    public function testSerialize(): void
+    {
+        $params = new URLSearchParams();
+        $params->append('a', 'b c');
+        self::assertSame($params->toString(), 'a=b+c');
+        $params->delete('a');
+        $params->append('a b', 'c');
+        self::assertSame($params->toString(), 'a+b=c');
+    }
+
+    /**
+     * @dataProvider provideSortingPayload
+     */
+    public function testSorting(string $input, array $output): void
+    {
+        if ('ﬃ&🌈' === $input) {
```

Comment:
@trowski This implementation is done with the following limitations:

Currently sorting is not done using codepoints as specified by the RFC. I tried several approaches but they all failed. (using `strcoll` and `setLocale` for instance) 

If you know a way to convert string into codepoints they I could "fix" this issue.

Reply:
@kelunik @trowski I improved the sorting algorithm with a rudimentary codepoint converter. The tests are all passing now 😂